cmd/agent/app/reporter/queue.go

```diff
@@ -0,0 +1,218 @@
+// Copyright (c) 2019 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reporter
+
+import (
+	"fmt"
+	"io"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/uber/jaeger-lib/metrics"
+	"go.uber.org/zap"
+
+	"github.com/jaegertracing/jaeger/cmd/agent/app/reporter/common"
+	"github.com/jaegertracing/jaeger/cmd/agent/app/reporter/queue"
+	"github.com/jaegertracing/jaeger/model"
+	jConverter "github.com/jaegertracing/jaeger/model/converter/thrift/jaeger"
+	"github.com/jaegertracing/jaeger/thrift-gen/jaeger"
+	"github.com/jaegertracing/jaeger/thrift-gen/zipkincore"
+)
+
+const (
+	defaultQueueType        = MEMORY
+	defaultBoundedQueueSize = 1000
+	defaultMaxRetryInterval = 20 * time.Second
+	defaultQueueWorkers     = 8
+)
+
+// Queue is generic interface which includes methods common to all implemented queues
+type Queue interface {
+	Enqueue(model.Batch) error
+	io.Closer
+}
+
+// QueuedReporter is a reporter that uses push-pull method that queues all incoming requests and then
+// lets the wrapped reporter do the actual pushing to the server (such as gRPC). If the requests fails
+// with retryable error the transaction is tried again.
+type QueuedReporter struct {
+	close uint32
+
+	wrapped Forwarder
+	queue   Queue
+	logger  *zap.Logger
+
+	retryMutex sync.Mutex
+
+	lastRetryIntervalChange time.Time
+	currentRetryInterval    time.Duration
+	maxRetryInterval        time.Duration
+	initialRetryInterval    time.Duration
+
+	reporterMetrics *MetricsReporter
+	agentTags       []model.KeyValue
+}
+
+// WrapWithQueue wraps the destination reporter with a queueing capabilities for retries
+func WrapWithQueue(opts *Options, forwarder Forwarder, logger *zap.Logger, mFactory metrics.Factory) *QueuedReporter {
+	q := &QueuedReporter{
+		wrapped:                 forwarder,
+		logger:                  logger,
+		lastRetryIntervalChange: time.Now(),
+		initialRetryInterval:    time.Millisecond * 100,
+		maxRetryInterval:        opts.ReporterMaxRetryInterval,
+		reporterMetrics:         NewMetricsReporter(forwarder, mFactory),
+		agentTags:               model.KeyValueFromMap(opts.AgentTags),
+		close:                   uint32(0),
+	}
+	q.currentRetryInterval = q.initialRetryInterval
+
+	// Some sanity checks
+	if opts.QueueType == "" {
```

Comment:
Does not seem needed? There can be default statement in the switch below.

Reply:
I thought it would make more sense to have it separated instead of switch's fallthrough. To me it makes different responsibilities of the code separated. Easier to read, nothing else. But YMMV.